stdlib/README.adoc

```diff
@@ -0,0 +1,102 @@
+= OCaml Stdlib
+
+== Design principles
+
+* When a module `M` has a main type it must be named `M.t`.
+* Functions should take the main type `t` as first argument (unless
+  there is a good reason for it, see next point).
+* If a functions takes another function as argument, that argument
+  should have the label `f` and not be last (to be able to omit the
+  label).
+* When a function may raise an exception as a non-exceptional
+  behavior, its name must have the suffix `_exn` to remind the user
+  that the exception should be handled.  Another function returning an
+  option instead must be provided.  Example:
+[source,ocaml]
+----
+val find : 'a t -> key -> 'a option
+val find_exn : 'a t -> key -> 'a
+----
+
+
+=== Collections
+
+* If a module `M` defines a collection of values, say `'a M.t`, it
+  must also define the following standard higher order functions:
+
+[source,ocaml]
+----
+val iter : f:('a -> unit) -> 'a t -> unit
+val iteri : f:(int -> 'a -> unit) -> 'a t -> unit
+val map : f:('a -> 'b) -> 'a t -> 'b t
+val mapi : f:(int -> 'a -> 'b) -> 'a t -> 'b t
+val fold : f:('a -> 'b -> 'a) -> init:'a -> 'b t -> 'a
+----
+
+The `iter`, `map` and `fold` function need not to specify the order in
+which arguments are passed to `f`.
+
+* Collections should be convertible to and from sequences by
+  providing:
+[source,ocaml]
+----
+val to_seq : 'a t -> 'a Seq.t
+val of_seq : 'a Seq.t -> 'a t
+----
+
+=== Arithmetic
+
+* When it makes sense, a module should provide _both_ standard infix
+  operators and named functions:
+
```

Comment:
I'm not sure why this is necessary as a requirement. It's not hard to use (+) -- the only difficult operator to use is `( * )`.

Reply:
The `add`,... are there for historical reasons.  Nowadays one may agree that infix operators are better. But maybe we would prefer to have `add`,... in the module and `( + )`,... in a, say, `Infix` or [`Ops`](https://github.com/ocaml/ocaml/pull/2097#issuecomment-430897682) sub-module?

Reply:
I think I agree with Chris' original notion here, both are a good idea.

Reply:
> The `add`,... are there for historical reasons.

I wouldn't say this. Remember that `M.()` comes with its own set of problems from a code evolution perspective.

Reply:
>  rather opposed to defining operators in the same modules as other identifiers,

Not sure how you envision it but this may also make writing code using binary operators/relations more cumbersome.  Maybe the module should export **two** sub-modules, one without binary operators/relations and one with them (and *also* with the same functions as the first sub-module).